plugins/mac-capture/mac-audio.m

```diff
@@ -637,6 +645,214 @@ static bool coreaudio_init(struct coreaudio_data *ca)
+	@"/Library/Audio/Plug-Ins/HAL/OBSDesktopAudioCapture.driver";
+
+static bool is_hal_device_updated()
+{
+	NSDictionary *halPlist =
+		[[NSBundle bundleWithPath:halPluginFileName] infoDictionary];
+	NSString *halVersion =
+		[halPlist objectForKey:@"CFBundleShortVersionString"];
+	NSString *halBuild =
+		[halPlist objectForKey:(NSString *)kCFBundleVersionKey];
+
+	NSString *obsVersion = [[[NSBundle mainBundle] infoDictionary]
+		objectForKey:@"CFBundleShortVersionString"];
+	NSString *obsBuild = [[[NSBundle mainBundle] infoDictionary]
+		objectForKey:(NSString *)kCFBundleVersionKey];
+	return [halVersion isEqualToString:obsVersion] &&
+	       [halBuild isEqualToString:obsBuild];
+}
+
+static bool is_hal_device_installed()
+{
+	AudioDeviceID deviceId;
+	coreaudio_get_device_id(CFSTR(DESKTOP_CAPTURE_HAL_DEVICE_UID),
+				&deviceId);
+	return deviceId != 0;
+}
+
+static bool modify_hal_plugin(bool reinstall)
+{
+	NSString *removeCmd;
+	NSString *installCmd;
+	removeCmd = [NSString
+		stringWithFormat:@"rm -rf '%@' && ", halPluginFileName];
+	if (reinstall) {
+		NSString *halPluginSourcePath = [[NSBundle mainBundle]
+			pathForResource:@"OBSDesktopAudioCapture.driver"
+				 ofType:nil];
+		NSString *halPluginDestinationPath =
+			@"/Library/Audio/Plug-Ins/HAL/";
+		installCmd =
+			[NSString stringWithFormat:@"cp -R '%@' '%@' && ",
+						   halPluginSourcePath,
+						   halPluginDestinationPath];
+	} else {
+		installCmd = @"";
+	}
+	NSString *restartCmd =
+		@"launchctl kickstart -k system/com.apple.audio.coreaudiod";
+	NSString *appleScriptString = [NSString
+		stringWithFormat:
+			@"do shell script \"%@%@%@\" with administrator privileges",
+			removeCmd, installCmd, restartCmd];
+
+	NSAppleScript *appleScriptObject =
+		[[NSAppleScript alloc] initWithSource:appleScriptString];
+	NSDictionary *error;
+	NSAppleEventDescriptor *eventDescriptor =
+		[appleScriptObject executeAndReturnError:&error];
+	if (eventDescriptor == nil) {
+		const char *errorMessage = [[error
+			objectForKey:@"NSAppleScriptErrorMessage"] UTF8String];
+		blog(LOG_WARNING,
+		     "[mac-capture]: Failed to modify HAL plugin: %s",
+		     errorMessage);
+		return false;
+	}
+
+	// Hack to wait until CoreAudio has reloaded. I hate it too.
+	sleep(5);
+
+	return true;
+}
+
+static bool is_multioutput_device_installed()
+{
+	AudioDeviceID deviceId;
+	coreaudio_get_device_id(CFSTR(DESKTOP_CAPTURE_MULTIOUTPUT_DEVICE_UID),
+				&deviceId);
+	return deviceId != 0;
+}
+
+static bool delete_multioutput_device()
+{
+	AudioDeviceID deviceId;
+	coreaudio_get_device_id(CFSTR(DESKTOP_CAPTURE_MULTIOUTPUT_DEVICE_UID),
+				&deviceId);
+
+	if (deviceId) {
+		/* Delete device */
+		OSStatus status = AudioHardwareDestroyAggregateDevice(deviceId);
+		if (status != noErr) {
+			blog(LOG_INFO, "Deleting old device failed. Reason: %d",
+			     status);
+			return false;
+		}
+		return true;
+	}
+	return false;
+}
+
+static inline bool get_default_monitoring_output_device(CFStringRef *cf_uid)
```

Comment:
Is this function misnamed maybe? It suggest that it retrieves the default monitoring output device, but it returns a boolean value instead.

Reply:
Upon second reading I realized this is a C API and the default device name is probably stored at the address the `CFStringRef` pointer refers to.